File: frontend/src/types/research.ts
// Research & Backtest types

export interface DatasetInfo {
  symbol: string;
  interval: string;
  bar_count: number;
  earliest: string | null;
  latest: string | null;
}

export interface DataSummary {
  datasets: DatasetInfo[];
  total_bars: number;
}

export interface CollectionStatus {
  job_id: string;
  exchange: string;
  status: string;
  symbols_total: number;
  symbols_done: number;
  bars_inserted: number;
  bars_skipped: number;
  errors: string[];
  progress_pct: number;
  started_at: string | null;
  completed_at: string | null;
}

export interface BacktestListItem {
  id: string;
  strategy_type: string;
  symbol: string;
  interval: string;
  status: string;
  total_trades: number;
  net_pnl: string;
  sharpe_ratio: string;
  max_drawdown_pct: string;
  win_rate: string;
  is_trustworthy: boolean;
  created_at: string;
}

export interface EquityCurvePoint {
  t: string;
  eq: string;
  dd: number;
}

export interface BacktestMetrics {
  strategy_name: string;
  symbol: string;
  period_start: string | null;
  period_end: string | null;
  total_net_pnl: string;
  total_gross_pnl: string;
  total_return_pct: number;
  sharpe_ratio: number;
  sortino_ratio: number;
  calmar_ratio: number;
  max_drawdown: string;
  max_drawdown_pct: number;
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  win_rate: number;
  profit_factor: number;
  expectancy: string;
  expectancy_ratio: number;
  avg_win: string;
  avg_loss: string;
  avg_holding_time_minutes: number;
  total_commission: string;
  total_slippage: string;
  fee_drag_pct: number;
  total_volume: string;
  trades_per_day: number;
  initial_capital: string;
  final_equity: string;
  capital_utilization_pct: number;
  is_trustworthy: boolean;
  trust_issues: string[];
  equity_curve: Array<{ timestamp: string; equity: string; drawdown_pct: number }>;
}

export interface BacktestDetail {
  id: string;
  strategy_type: string;
  symbol: string;
  interval: string;
  status: string;
  strategy_params: Record<string, unknown>;
  cost_model: string;
  initial_capital: string;
  total_trades: number;
  total_bars: number;
  metrics: BacktestMetrics | null;
  equity_curve: EquityCurvePoint[] | null;
  is_trustworthy: boolean;
  trust_issues: string[] | null;
  error: string | null;
  created_at: string;
}

export interface SweepResultItem {
  params: Record<string, unknown>;
  sharpe_ratio: number;
  net_pnl: string;
  max_drawdown_pct: number;
  win_rate: number;
  total_trades: number;
  profit_factor: number;
  rank_score: number;
  is_trustworthy: boolean;
}

export interface ResearchDashboard {
  data_summary: DataSummary;
  active_strategies: number;
  total_backtests: number;
  best_sharpe: { strategy_type: string; symbol: string; sharpe: string; net_pnl: string } | null;
  worst_drawdown: { strategy_type: string; symbol: string; drawdown: string } | null;
  recent_backtests: BacktestListItem[];
  live_strategy_pnl: Record<string, string>;
}

export interface StrategyEngineItem {
  id: string;
  name: string;
  description: string | null;
  strategy_type: string;
  status: string;
  trading_mode: string;
  config_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface StrategyRuntimeStatus {
  name: string;
  status: string;
  strategy_type: string;
  uptime_seconds: number | null;
  ticks_processed: number;
  orders_submitted: number;
  orders_filled: number;
  active_orders: number;
  current_inventory: Record<string, string>;
  pnl: {
    strategy_name: string;
    realized_pnl: string;
    unrealized_pnl: string;
    net_pnl: string;
    total_trades: number;
    winning_trades: number;
    losing_trades: number;
    win_rate: number;
    total_volume: string;
    total_commission: string;
    pnl_per_trade: string;
    start_time: string;
  };
}

File: backend/app/api/schemas/research.py
"""
Request/response schemas for the research & backtest API.
"""

from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Any, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Data Collection
# ---------------------------------------------------------------------------

class DataCollectionRequest(BaseModel):
    exchange: str = Field("yfinance", description="Data source id")
    symbols: list[str] = Field(
        default=["EURUSD", "GBPUSD", "USDJPY", "AAPL", "MSFT", "SPY"],
        description="Symbols to collect (forex pairs or stock tickers)",
    )
    intervals: list[str] = Field(
        default=["1h", "1d"],
        description="Candle intervals",
    )
    limit: int = Field(500, ge=1, le=1000, description="Candles per request")


class DataCollectionStatus(BaseModel):
    job_id: str
    exchange: str
    status: str
    symbols_total: int
    symbols_done: int
    bars_inserted: int
    bars_skipped: int
    errors: list[str]
    progress_pct: float
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None


class DataSummaryResponse(BaseModel):
    datasets: list[dict]
    total_bars: int


# ---------------------------------------------------------------------------
# Backtesting
# ---------------------------------------------------------------------------

class BacktestRequest(BaseModel):
    strategy_type: str = Field(
        ...,
        description="Strategy type: grid, mean_reversion, market_making, breakout, sma_crossover, rsi, bollinger, macd",
    )
    symbol: str = Field("EURUSD", description="Trading symbol")
    interval: str = Field("5m", description="Candle interval to backtest on")
    initial_capital: Decimal = Field(
        Decimal("10000"), ge=0, description="Starting capital"
    )
    cost_model: str = Field(
        "forex",
        description="Cost model: forex, forex_ecn, stock, stock_ib, conservative, zero",
    )
    strategy_params: dict[str, Any] = Field(
        default_factory=dict,
        description="Strategy-specific parameters",
    )
    max_position_size: Decimal = Field(
        Decimal("10000"), description="Max position size"
    )
    stop_loss_pct: Optional[float] = Field(
        5.0, description="Stop loss % (null to disable)"
    )
    take_profit_pct: Optional[float] = Field(
        None, description="Take profit % (null to disable)"
    )
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None


class BacktestResponse(BaseModel):
    id: str
    status: str
    total_bars: int
    total_trades: int
    run_at: datetime
    config: dict
    metrics: Optional[dict] = None
    error: Optional[str] = None


class BacktestListItem(BaseModel):
    id: str
    strategy_type: str
    symbol: str
    interval: str
    status: str
    total_trades: int
    net_pnl: str
    sharpe_ratio: str
    max_drawdown_pct: str
    win_rate: str
    is_trustworthy: bool
    created_at: datetime


# ---------------------------------------------------------------------------
# Parameter Sweep
# ---------------------------------------------------------------------------

class ParameterSweepRequest(BaseModel):
    strategy_type: str
    symbol: str = "EURUSD"
    interval: str = "5m"
    initial_capital: Decimal = Decimal("10000")
    param_grid: dict[str, list] = Field(
        ...,
        description="Dict of param_name -> list of values to sweep",
    )
    max_position_size: Decimal = Decimal("10000")


class SweepResultItem(BaseModel):
    params: dict
    sharpe_ratio: float
    net_pnl: str
    max_drawdown_pct: float
    win_rate: float
    total_trades: int
    profit_factor: float
    rank_score: float
    is_trustworthy: bool


class ParameterSweepResponse(BaseModel):
    strategy_type: str
    symbol: str
    total_combinations: int
    results: list[SweepResultItem]


# ---------------------------------------------------------------------------
# Metrics
# ---------------------------------------------------------------------------

class EquityCurvePoint(BaseModel):
    timestamp: datetime
    equity: str
    drawdown_pct: float


class StrategyMetricsResponse(BaseModel):
    strategy_name: str
    symbol: str
    period_start: Optional[datetime] = None
    period_end: Optional[datetime] = None
    total_net_pnl: str
    total_return_pct: float
    sharpe_ratio: float
    sortino_ratio: float
    max_drawdown_pct: float
    total_trades: int
    win_rate: float
    profit_factor: float
    expectancy: str
    avg_holding_time_minutes: float
    total_commission: str
    fee_drag_pct: float
    trades_per_day: float
    is_trustworthy: bool
    trust_issues: list[str]
    equity_curve: list[EquityCurvePoint]


# ---------------------------------------------------------------------------
# Research Dashboard
# ---------------------------------------------------------------------------

class ResearchDashboardResponse(BaseModel):
    """Aggregated dashboard data for the quant research view."""

    # Data status
    data_summary: DataSummaryResponse

    # Strategy performance overview
    active_strategies: int = 0
    total_backtests: int = 0
    best_sharpe: Optional[dict] = None
    worst_drawdown: Optional[dict] = None

    # Recent activity
    recent_backtests: list[BacktestListItem] = []
    live_strategy_pnl: dict[str, str] = {}
